sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/codegen/CodeGenerator.scala

```diff
@@ -697,7 +697,14 @@ class CodegenContext {
           }
         """
       s"${addNewFunction(compareFunc, funcCode)}($c1, $c2)"
-    case other if other.isInstanceOf[AtomicType] => s"$c1.compare($c2)"
+    case other if other.isInstanceOf[AtomicType] =>
+      s"""
+        if ($c1.getClass.getMethods.map(_.getName).filter(_ matches "compare").size == 1) {
+          $c1.compare($c2)
```

Comment:
hmm, `genComp` seems to return a java statement evaluated to boolean, are you sure returning a block like this work?

Reply:
+1